### PR TITLE
Reduce MiniYaml allocations

### PR DESCRIPTION
**Pull request by RoosterDragon**
Reduces MiniYaml allocations during:
- `MiniYaml.Merge`: Various collections are pre-sized to a sensible capacity in advance, to avoid reallocating a larger capacity during various operations.
- `MiniYaml.FromLines`: A new extension method allows iterating over the lines from a stream as `ReadOnlyMemory`, avoiding the need to allocate a string for every line in the stream. The parser can then use span operations to avoid realising intermediate strings. Only the final strings required to populate the output data structure are required to be allocated.

This results in some decent reductions in allocations without having to touch the core logic.
`MiniYaml.Merge` now allocates 88% compared to original. `MiniYaml.FromLines` now allocates 72% compared to original.

### CODE REVIEW

File: OpenRA.Game/MiniYaml.cs [hunk 1]
```diff
@@ -241,40 +242,43 @@ static List<MiniYamlNode> FromLines(IEnumerable<string> lines, string filename,
...
+						value = value.ToString().Replace("\\#", "#");
 				}
 
 				if (key != null || !discardCommentsAndWhitespace)
 				{
-					key = key == null ? null : stringPool.GetOrAdd(key, key);
-					value = value == null ? null : stringPool.GetOrAdd(value, value);
-					comment = comment == null ? null : stringPool.GetOrAdd(comment, comment);
+					var keyString = key == null ? null : key.ToString();
+					var valueString = value == null ? null : value.ToString();
```
Comment (pchote) [inline OpenRA.Game/MiniYaml.cs]: You've removed the `value != null` checks above, suggesting that it shouldn't be needed here either?

File: OpenRA.Test/OpenRA.Game/StreamExtsTests.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿using System.IO;
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/StreamExtsTests.cs]: Can we please add the standard file header here?

**Review by pchote — COMMENTED** [review]
Just a couple of small nits, otherwise looks reasonable to me.

File: OpenRA.Game/MiniYaml.cs (hunk 1 above)
Comment (RoosterDragon) [reply OpenRA.Game/MiniYaml.cs]: `ReadOnlySpan<char> key = null;` produces an empty span. i.e. `key.IsEmpty` is true and `key.Length` gives 0.

Therefore we can replace the null checks with not-empty checks - but the existing logic makes those not-empty checks redundant so they can go away. We need a not-empty check later to avoid allocating empty strings though.

Perhaps this would be clearer by assigning `ReadOnlySpan<char> key = default;` and using `var keyString = key.IsEmpty ? null : key.ToString();` instead?

Comment (pchote) [reply OpenRA.Game/MiniYaml.cs]: Yeah, that would be much clearer 👍

Comment (RoosterDragon) [reply OpenRA.Game/MiniYaml.cs]: Added. Also added some test cases to cover the comment field being empty when expected as that requires `comment == default` to work as desired.

File: OpenRA.Game/MiniYaml.cs [hunk 3]
```diff
@@ -254,25 +254,25 @@ static List<MiniYamlNode> FromLines(IEnumerable<ReadOnlyMemory<char>> lines, str
...
 
-				if (key != null || !discardCommentsAndWhitespace)
+				if (!key.IsEmpty || !discardCommentsAndWhitespace)
 				{
-					var keyString = key == null ? null : key.ToString();
-					var valueString = value == null ? null : value.ToString();
-					var commentString = comment == null ? null : comment.ToString();
+					var keyString = key.IsEmpty ? null : key.ToString();
+					var valueString = value.IsEmpty ? null : value.ToString();
+					var commentString = comment == default ? null : comment.ToString();
```
Comment (pchote) [inline OpenRA.Game/MiniYaml.cs:278]: Can we please add a comment like
```csharp
// Note: We need to support empty comments here to ensure that empty comments
// (i.e. a lone # at the end of a line) can be correctly re-serialized
```

here? The need for this difference wasn't obvious to me until I spent some time poking at the new testcase.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b79cc451b203c0437950315a9c7f118c6d34fc7d...c17481e911fb0abe0a47009656841d1db4898f3a)
